collector/lib/HostInfo.cpp

```diff
@@ -179,6 +182,75 @@ bool HostInfo::HasEBPFSupport() {
   return collector::hasEBPFSupport(kernel, GetOSID());
 }
 
+bool HostInfo::HasBTFKernelSupport() {
+  std::string btf_vmlinux_path = GetHostPath("/sys/kernel/btf/vmlinux");
+
+  if (faccessat(AT_FDCWD, btf_vmlinux_path.c_str(), R_OK, AT_EACCESS) == -1) {
+    if (errno == ENOTDIR || errno == ENOENT) {
+      CLOG(DEBUG) << "BTF debug symbols file not found in sysfs";
+    } else {
+      CLOG(WARNING) << "Unable to access " << btf_vmlinux_path << ": " << StrError();
+    }
+    return false;
+  }
+
+  CLOG(DEBUG) << "BTF debug symbols file found. Kernel BTF support available.";
+
+  return true;
+}
+
+bool HostInfo::HasBTFSymbols() {
+  const char* locations[] = {
+      /* try canonical vmlinux BTF through sysfs first */
+      "/sys/kernel/btf/vmlinux",
+      /* fall back to trying to find vmlinux on disk otherwise */
+      "/boot/vmlinux-%1$s",
+      "/lib/modules/%1$s/vmlinux-%1$s",
+      "/lib/modules/%1$s/build/vmlinux",
+      "/usr/lib/modules/%1$s/kernel/vmlinux",
+      "/usr/lib/debug/boot/vmlinux-%1$s",
+      "/usr/lib/debug/boot/vmlinux-%1$s.debug",
+      "/usr/lib/debug/lib/modules/%1$s/vmlinux",
+      0};
+
+  char path[PATH_MAX + 1];
+  const char* const* location;
+
+  for (location = locations; *location; location++) {
+    snprintf(path, PATH_MAX, *location, kernel_version_.release.c_str());
+    std::string host_path = GetHostPath(path);
+
+    if (faccessat(AT_FDCWD, host_path.c_str(), R_OK, AT_EACCESS) == 0) {
+      CLOG(DEBUG) << "BTF symbols found in " << host_path;
+      return true;
+
+    } else {
+      if (errno == ENOTDIR || errno == ENOENT) {
+        CLOG(DEBUG) << host_path << " does not exist";
+      } else {
+        CLOG(WARNING) << "Unable to access " << host_path << ": " << StrError();
+      }
+    }
+  }
+
+  CLOG(DEBUG) << "Unable to find BTF symbols in any of the usual locations.";
+
+  return false;
+}
+
+bool HostInfo::HasBPFRingBufferSupport() {
+  int res;
+
+  res = libbpf_probe_bpf_map_type(BPF_MAP_TYPE_RINGBUF, NULL);
+
+  if (res < 0) {
+    CLOG(WARNING) << "Unable to check for the BPF RingBuffer availability. "
```

Comment:
In this case `libbpf` says feature detection may failed, makes sense to report `errno`?

Reply:
I couldn't find a reference to the errno availability in the docs for that case. And the code suggests that it does not reflect the actual error reason.

Reply:
This one [[1]]? It seems this function does `sys_bpf_fd` and returns `-errno` if `fd < 0`.

[1]: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/tools/lib/bpf/libbpf.h#n1371

Reply:
errno is tested underneath, but when the error if reported at the API level, I am afraid this information is lost.